### DISCUSSION TITLE
Label Text selectable

### DISCUSSION THREAD

**Discussion by charleschege** [emilk/egui#353] [Q&A]
Is there a way to make label text selectable because currently it is not?

**Comment by emilk** [comment]
No. I believe a "Click to copy" button is way better UX anyway:

``` rust
if ui.button("📋").on_hover_text("Click to copy").clicked() {
    ui.output().copied_text = what_the_user_is_interested_in;
}
```

**Reply by charleschege** [reply]
What if a user wants to select only one sentence or word in a web page article? Offering a copy option on the whole Web page would not make sense

**Reply by emilk** [reply]
Yeah, for that case we should support text selection.

It just hasn't been a priority, since so far each time someone has wanted the feature they have wanted it for the case that is better covered with the "Click to copy" solution :)

**Reply by gitmalong** [reply]
I think being able to select and copy text belongs to the basic feature set people expect from a web app.

**Comment by dimvoly** [comment]
In case it helps anyone, clickable label:

```rust
if ui
    .add(
        egui::Label::new("Label Text")
            .sense(egui::Sense::click()),
    )
    .on_hover_text("Click to copy 'Label Text'")
    .clicked()
{
    ui.output_mut(|po| {
        po.copied_text = String::from("Label Text");
    });
}
```
I'm on egui 0.21.0.

**Comment by YgorSouza** [comment]
For the record, this is now possible with the current egui master. I'm not sure of when it was introduced. Simply pass a `&str` to the `TextEdit` widget and it will be rendered like a `Label`, but selectable and copyable. Note that you have to make it multiline in order to display line breaks properly, and then it will have the same minimum size as the regular multiline `TextEdit`, unless you change it. It also has a bit more padding than the `Label` in all directions, and a slightly different default color, so that's something to watch out for.

```rs
ui.label("Label for comparison");
ui.add(
    TextEdit::multiline(&mut "Immutable &str\nSecond line")
        .min_size([0.0, 0.0].into())
        .desired_rows(1),
);
let mut mut_string = String::from("Mutable string");
ui.text_edit_multiline(&mut mut_string.as_str());
ui.text_edit_multiline(&mut mut_string);

```

![image](https://github.com/emilk/egui/assets/43298013/087f584c-6145-4caf-83f2-158393c9dd5b)

**Reply by liangzid** [reply]
Unfortunately, this kind of method is only suitable to `Text` and not available for `WidgetText` like `LayeroutJob`, `RichText`, and so on.

I meet this requirement just because I'd like to write a document viewer with highlight function like this: 
![图片](https://github.com/emilk/egui/assets/26396476/7b183793-1706-4f88-9389-f5c079c53c45)

**Reply by YgorSouza** [reply]
I've never done this, but maybe the EasyMark editor demo is a good place to start.

https://github.com/emilk/egui/blob/b5e350206758bd1b89d7dea26edeb024b4cf6329/crates/egui_demo_lib/src/easy_mark/easy_mark_editor.rs#L75-L95

If I change line 88 to `egui::TextEdit::multiline(&mut code.as_str())`, it is rendered as a label and not editable, but is still selectable and has all the formatting.

![image](https://github.com/emilk/egui/assets/43298013/b7b4e0e8-858e-43f0-81e0-ae9d7a15b519)

**Reply by liangzid** [reply]
Great reference. Thank you, I do it!
